sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+
+import scala.collection.mutable.ArrayBuffer
+
+import com.google.common.io.ByteStreams
+
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.array.ByteArrayMethods
+import org.apache.spark.unsafe.memory.MemoryBlock
+import org.apache.spark.{TaskContext, SparkEnv}
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+
+/**
+ * ::Experimental::
+ *
+ * @since 1.6.0
+ */
+@Experimental
+object TungstenCache {
+  def cache(
+      child: SparkPlan,
+      compressionType: Option[String] = None,
+      blockSize: Int = 4 * 1000 * 1000): RDD[InternalRow] = {
+    val numFields = child.schema.length
+    val childRDD: RDD[UnsafeRow] = {
+      if (child.outputsUnsafeRows) {
+        child.execute().asInstanceOf[RDD[UnsafeRow]]
+      } else {
+        ConvertToUnsafe(child).execute().asInstanceOf[RDD[UnsafeRow]]
+      }
+    }
+    val cachedRDD: RDD[MemoryBlock] = childRDD.mapPartitions { rowIterator =>
+      val bufferedRowIterator = rowIterator.buffered
+      val taskMemoryManager = TaskContext.get().taskMemoryManager()
```

Comment:
I'm wondering what we actually gain by using the TaskMemoryManager here. The TMM would allow us to cache off-heap but it looks like we don't / won't currently do that if we use compression. For the initial version of this task, I'd like to focus on making the basic caching work really well and defer the question of how to integrate offheap caching with block manager to a followup PR.